src/dns/mod.rs

```diff
@@ -3,44 +3,107 @@ use std::net::{
     SocketAddr,
     SocketAddrV4, SocketAddrV6,
 };
-use std::vec;
 use c_ares::AResults;
 
 mod c_ares;
 
 pub use self::c_ares::GLOBAL_RESOLVER;
 
 pub struct IpAddrs {
-    iter: vec::IntoIter<SocketAddr>,
+    inner: SingleOrShifted<SocketAddr>,
```

Comment:
It seems like a lot of the logic here could be simplified..

```rust
pub struct IpAddrs {
    offset: usize,
    current: usize,
    data: Vec<SocketAddr>
}

impl Iterator for IpAddrs {
    type Item = SocketAddr;
    
    fn next(&mut self) -> Option<Self::Item> {
        if current == self.data.len() {
            return None;
        }
        
        let index = (self.current + self.offset) % self.data.len();
        let item = *self.data[index];
        self.current += 1;
        Some(item)
    }
}
```

No enum needed, no option dancing, and fewer changes.

Reply:
I use the enum to avoid creating a Vec when only one item is needed, do we want to add that back?

The other thing is wouldn't we require a `Vec<Option<T>>` so that we can `take()` and return it with ownership as `SocketAddr` is not `Copy`?  I think `let item = *self.data[index];` wouldn't compile because it's trying taking ownership of `SocketAddr`.

Reply:
> I use the enum to avoid creating a Vec when only one item is needed, do we want to add that back?

This isn't a big deal imo. In most cases, we'll have more than one entry anyway. If we don't, maybe there's something funny with c-ares. Plus, it's good to value simplicity. Way easier to work on code that's approachable.

> The other thing is wouldn't we require a Vec<Option<T>> so that we can take() and return it with ownership as SocketAddr is not Copy? I think let item = *self.data[index]; wouldn't compile because it's trying taking ownership of SocketAddr.

`SocketAddr` is `Copy`. I checked the docs before suggesting this. Even if it's not, a clone wouldn't be a big deal.